### PR TITLE
[metasrv] feature: exchange protocol version with client

### PR DESCRIPTION
**Pull request by drmingdrmer**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

##### [metasrv] feature: exchange protocol version with client
Before transfer any data, `MetaGrpcClient` must ensure it is compatible
with the metasrv it is connecting to, in the handshake RPC.

Add compatibility cross test:

```
latest query -- latest metasrv
old    query -- latest metasrv
latest query -- old    metasrv
```

Features

- Compatibility test: download test suite for query.

  When testing an old query with latest metasrv,
  download the test suite for that version of query instead of using
  the latest suite, which may not pass if new feature test is added
  into the latest suite but not yet impl-ed in the old query.

- Compatibility test:

  Test the compatibility between:
  - The latest query and the latest metasrv.
  - The oldest compatible query with latest metasrv.
  - The latest query with oldest compatible  metasrv.

- Protocol version exchange when meta-client `handshake` with metasrv.

  meta-client(query) keeps track of the min-metasrv version it is compatible
  with.

  metasrv keeps track of the min-meta-client version it is compatible
  with.

  Client handshake check these versions to ensure compatibility.
  Otherwise an error is returned and no further data will be transferred.

- Add admin CLI command `--cmd ver`.

  `databend-query --cmd ver` outputs its build version and the minimal
  compatible metasrv version, such as:

  ```
  version: 0.7.61-nightly
  min-compatible-metasrv-version: 0.7.59
  ```

  `databend-meta --cmd ver` outputs its build version and the minimal
  compatible meta-client(query) version, such as:

  ```
  version: 0.7.61-nightly
  min-compatible-client-version: 0.7.57
  ```

Refactors

- Add a new error `InvalidArgument` to indicate protocol-error

- `MetaGrpcClient::try_new()` does not need to be `async`.

- Metasrv config does not `check()` if `--cmd` is non-empty:

  When `cmd` is specified, it does not mean to run it.

Docs

- Add doc explain how to check versions and how to upgrade.

---

- Fix: #5627
- Fix: #4948

## Changelog

- New Feature

## Related Issues

### CODE REVIEW

**Comment by BohuTANG** [comment]
Cool.
If we build from source with out git information, what happenes during metasrv and query?

**Review by lichuang — APPROVED** [review]

**Comment by drmingdrmer** [comment]
> Cool. If we build from source with out git information, what happenes during metasrv and query?

The commit-version will be missing and the query won't be able to connect to metasrv.

**Review by BohuTANG — APPROVED** [review]

**Comment by drmingdrmer** [comment]
> Cool. If we build from source with out git information, what happenes during metasrv and query?

BTW I've once considered using the version defined in `Cargo.toml` for building a release.

```
[package]
name = "databend-query"
version = "0.7.0"
```

E.g., if the version in Cargo is `0.7.0`, then a nightly build version could be `0.7.0+20220528`.

Then the git-tag triggered release action only is used for formal releases.

**Comment by BohuTANG** [comment]
> > Cool. If we build from source with out git information, what happenes during metasrv and query?
> 
> The commit-version will be missing and the query won't be able to connect to metasrv.

I think this is OK.

**Comment by drmingdrmer** [comment]
> > > Cool. If we build from source with out git information, what happenes during metasrv and query?
> > 
> > 
> > The commit-version will be missing and the query won't be able to connect to metasrv.
> 
> I think this is OK.

Hmm... looks like CI failed due to this issue. In the CI it just checkout one last commit without any tags. The version can not be determined.

It can be worked around by using a special protocol-version == 0, which I do not like. 🤔

**Comment by BohuTANG** [comment]
> > > > Cool. If we build from source with out git information, what happenes during metasrv and query?
> > > 
> > > 
> > > The commit-version will be missing and the query won't be able to connect to metasrv.
> > 
> > 
> > I think this is OK.
> 
> Hmm... looks like CI failed due to this issue. In the CI it just checkout one last commit without any tags. The version can not be determined.
> 
> It can be worked around by using a special protocol-version == 0, which I do not like. 🤔

@everpcpc fixed it, may help here

File: .github/actions/test_compat/action.yml [hunk 1]
```diff
@@ -15,6 +15,10 @@ runs:
     - name: Maximize build space
       uses: ./.github/actions/cleanup
 
+    - name: Rust setup
```
Comment (everpcpc) [inline .github/actions/test_compat/action.yml:18]: `dev_setup` is only needed on macos.
on linux ci,  `setup_build_tool` already made it working.

**Comment by everpcpc** [comment]
> > > > > Cool. If we build from source with out git information, what happenes during metasrv and query?
> > > > 
> > > > 
> > > > The commit-version will be missing and the query won't be able to connect to metasrv.
> > > 
> > > 
> > > I think this is OK.
> > 
> > 
> > Hmm... looks like CI failed due to this issue. In the CI it just checkout one last commit without any tags. The version can not be determined.
> > It can be worked around by using a special protocol-version == 0, which I do not like. 🤔
> 
> @everpcpc fixed it, may help here

should be fixed with: https://github.com/datafuselabs/databend/pull/5520/files#r884142502

File: .github/actions/test_compat/action.yml (hunk 1 above)
Comment (drmingdrmer) [reply .github/actions/test_compat/action.yml:18]: > > @everpcpc fixed it, may help here
> 
> should be fixed with: https://github.com/datafuselabs/databend/pull/5520/files#r884142502

Is it appropriate to fetch the full history? 
The fetching time would be longer then.

**Comment by drmingdrmer** [comment]
> should be fixed with: https://github.com/datafuselabs/databend/pull/5520/files#r884142502

Is it appropriate to fetch the full history?
The fetching time would be longer then.

File: .github/actions/test_compat/action.yml [hunk 2]
```diff
@@ -15,6 +15,10 @@ runs:
...
+      run: |
+        bash ./scripts/setup/dev_setup.sh -yb
+
     - name: Setup Build Tool
       uses: ./.github/actions/setup_build_tool
```
Comment (everpcpc) [inline .github/actions/test_compat/action.yml:24]: installing python libs is also not need, and already done by build tool image

File: .github/actions/test_compat/action.yml (hunk 1 above)
Comment (drmingdrmer) [reply .github/actions/test_compat/action.yml:18]: It complains in `test-compat.sh` that `wget` is not found.
What did I do wrongly?

I was trying to run it with build-tool:
`build-tool bash ./tests/compat/test-compat.sh`

`wget` should be already provided by github-action.

**Comment by everpcpc** [comment]
> 

we are already doing this with release build, and it seems checkout takes around 10s which seems fine.

**Comment by everpcpc** [comment]
> > should be fixed with: https://github.com/datafuselabs/databend/pull/5520/files#r884142502
> 
> Is it appropriate to fetch the full history? The fetching time would be longer then.

we are already doing this with release build, and the checkout takes around 10s which seems fine.

File: .github/actions/test_compat/action.yml (hunk 1 above)
Comment (everpcpc) [reply .github/actions/test_compat/action.yml:18]: ah, it's not in the build-tool image now. 
maybe you can use curl instead.

Comment (drmingdrmer) [reply .github/actions/test_compat/action.yml:18]: That's all right! Thanks!

**Pull request merged**